marginalia.el

```diff
@@ -799,6 +827,38 @@ These annotations are skipped for remote paths."
         (push attrs marginalia--fontified-file-attributes)
         attrs)))
 
+(defun marginalia--relative-age (time)
+  "Format TIME as a relative age."
+  (string-pad
+   (replace-regexp-in-string
+    "\\([0-9]+\\)\\.[0-9]+\\([a-z]\\)"
+    (lambda (age)
+      (concat (match-string 1 age) " "
+              (or (cdr (assoc (match-string 2 age) marginalia--relative-age-units))
+                  (match-string 2 age))
+              (unless (string= "1" (match-string 1 age))
+                "s")
+              " ago"))
+    (seconds-to-string (float-time (time-since time))))
+   12))
+
+(defun marginalia--absolute-age (time)
+  "Format TIME as an absolute age."
+  (format-time-string
+   (if (> (decoded-time-year (decode-time (current-time)))
+          (decoded-time-year (decode-time time)))
+       "%Y %b %d "
+     "%b %d %H:%M")
+   time))
+
+(defun marginalia--file-age (time)
+  "Format file age TIME, suitably for use in annotations."
+  (propertize
+   (if (or (null margialia-max-relative-age) (< (float-time (time-since time)) margialia-max-relative-age))
```

Comment:
What about removing the indirection here? Offer three functions marginalia-date-absolute, marginalia-date-relative, marginalia-date-mixed and then a variable marginalia-date-formatter.

Reply:
I like the sound of this. The one downside I can see is that if you want to change the absolute format that `-mixed` uses you'd need to either:
- add customisation variables for `-absolute`
- advise `-absolute`
- write a new `-mixed` function

None of which sound great to me.

Reply:
My idea is to add keyword arguments to each of these functions which allow customization. Then the formatter variable can be a list like this `(marginalia-date-relative :option1 ... :option2 ...)`

Reply:
Hmmm, so basically relocating the complexity from variables to arguments. Not sure how I feel about this, for example with `marginalia--absolute-age` being so simple, I'd imagine writing your own variant function to be easier than working out the right arguments and how to apply them.

```lisp
(defun marginalia--absolute-age (time)
  "Format TIME as an absolute age."
  (format-time-string
   (if (> (decoded-time-year (decode-time (current-time)))
          (decoded-time-year (decode-time time)))
       "%Y %b %d "
     "%b %d %H:%M")
   time))
```

Reply:
That said, my comment above doesn't really apply to the `-relative-age` function, but having a func var for one but not the other feels inconsistent to me.

Reply:
@minad just because I know you're online I'm going to try a gentle ping -- I feel like we've almost got this worked out :smiley:

Reply:
Once again, a thought has occured. I'm going to try changing some things around in line with your last comment.